### ISSUE TITLE
Docker Trojan Horse in Redis Container

### ISSUE THREAD

**Issue by zimo-xiao** [laradock/laradock#2254]
### Info:
- Docker version (`$ docker --version`): 
- Laradock commit (`$ git rev-parse HEAD`): 
- System info (Mac, PC, Linux): 
- System info disto/version: 

### Issue:
I recently found that after a rebuild, the redis container would shut down after exactly 24 hr. After looking through the log, I found that after 24hr, redis would run the following command:

redis_1                | sh: 1: killall: not found
redis_1                | ./xmrig-notls: unrecognized option '--max-cpu-usage'
_____

### Expected behavior:
Expected it to work
_____

### Reproduce:
That's the tricky part. I didn't have this kind of issue until last week. Maybe the horse is from official docker redis container (possibly not) or maybe an intrusion into my system, but I only found xmrig under overlay2:

/var/lib/docker/overlay2/f0d664c7954a158354876e11b8c89bd3c323d49ddc59c5531750b5ac639f9c6e/merged/tmp/xmrig-2.99.0-beta/xmrig
/var/lib/docker/overlay2/f0d664c7954a158354876e11b8c89bd3c323d49ddc59c5531750b5ac639f9c6e/diff/tmp/xmrig-2.99.0-beta/xmrig

So it's more likely a container problem. 
_____

### Relevant Code:

```
// place a code sample here
```

**Comment by cwilby** [comment]
@zimo-xiao This happened because you didn't secure your Redis container (even though you think you did!) I know because this exact situation happened to me before :o)

If your docker-compose.yml file contains this redis configuration:

```yml
services:
  redis:
    ..
    ports:
      - 6379:6379
```

.. then Docker will override your iptables.  This doesn't show in `ufw status` because it wasn't configured there.

This situation is explained in more detail [here](https://fralef.me/docker-and-iptables.html).

The solution is to change `6379:6379` to `127.0.0.1:6379:6379`, and create a secure password for Redis.

**Comment by cwilby** [comment]
Additionally, Laradock doesn't have an env option for setting the redis password, so you will need to   run sed on the redis.conf file.

In `.env`:
```
REDIS_PASSWORD=supersecret
```

In `./laradock/redis/redis.conf`:
```
requirepass __REDIS_PASSWORD__
```

In `./laradock/redis/Dockerfile`

```
FROM redis:latest

ARG REDIS_PASSWORD=secret

RUN mkdir -p /usr/local/etc/redis
COPY redis.conf /usr/local/etc/redis/redis.conf
RUN sed -i "s/__REDIS_PASSWORD__/${REDIS_PASSWORD}/g" /usr/local/etc/redis/redis.conf

VOLUME /data

EXPOSE 6379

CMD ["redis-server", "/usr/local/etc/redis/redis.conf"]
```

Darn cryptojackers.

**Comment by javierrivarola** [comment]
I'm having the same issue, after trying to set a password, laravel won't connect to redis no matter what, is there any step missing from what is mentioned above?

**Comment by cwilby** [comment]
@javierrivarola Couple of things to check:

1. Are you able to connect to Redis via redis-cli, type `auth`, enter password, and run `ping`?
```shell
$ docker-compose exec redis redis-cli
127.0.0.1:6379> auth {{your_password}}
127.0.0.1:6379> ping
pong
```

2. Check that you have set your `REDIS_PASSWORD` in the .env file used by Laravel.

3. Check `config/database.php` and make sure that the default redis driver includes a `password` field that pulls from `env('REDIS_PASSWORD', null)`.

Failing that, it could be a build-cache issue, where you've set the files but it hasn't updated in the container.

```shell
$ docker-compose down && docker-compose up -d --build
```

**Comment by javierrivarola** [comment]
Yes i can auth from the redis-cli but no matter what i do, laravel gets "Connection refused [tcp://redis:6379]", already tried to re build workspace container but nothing

**Comment by cwilby** [comment]
I also get that error message when starting my containers for the first time for about 10-30 seconds while Redis is getting started.

After all containers are running, check to see if this works:

```shell
$ docker-compose exec workspace bash
$ art tinker
> cache(['foo' => 'bar'], 10);
> cache('foo')
```

**Comment by javierrivarola** [comment]
Laravel just won't connect 
<img width="1211" alt="Screen Shot 2019-09-03 at 1 04 02 PM" src="https://user-images.githubusercontent.com/7185059/64193747-6bebdf00-ce4b-11e9-8b48-b7e0254306b6.png">

my .env
<img width="222" alt="Screen Shot 2019-09-03 at 1 04 38 PM" src="https://user-images.githubusercontent.com/7185059/64193754-71492980-ce4b-11e9-8a99-d534864b1534.png">

**Comment by cwilby** [comment]
Can you post the `redis` service from your `docker-compose.yml` file?

**Comment by javierrivarola** [comment]
<img width="431" alt="Screen Shot 2019-09-03 at 1 07 56 PM" src="https://user-images.githubusercontent.com/7185059/64193928-ec124480-ce4b-11e9-987e-d168de9e24b0.png">

and this is my laradock .env
<img width="477" alt="Screen Shot 2019-09-03 at 1 08 16 PM" src="https://user-images.githubusercontent.com/7185059/64193934-f2082580-ce4b-11e9-82cb-45f4a7ef869e.png">

**Comment by cwilby** [comment]
Thanks @javierrivarola, I've updated my answer above to include adding `REDIS_PASSWORD` as a build argument, which was missing in your service.

**Comment by javierrivarola** [comment]
i'm not seeing any edited answer about adding `REDIS_PASSWORD` in `./laradock/docker-compose.yml`, can you clarify?

**Comment by javierrivarola** [comment]
do i need to rebuild everything? still not working

**Comment by cwilby** [comment]
Yep, need to rebuild so that the argument gets used.  You can run `docker-compose down` then `docker-compose up -d --build` to trigger a build with only new changes.

**Comment by javierrivarola** [comment]
The same issue after rebuilding everything :S

**Comment by cwilby** [comment]
We got this :)

Can you post `./laradock/redis/Dockerfile`?

**Comment by javierrivarola** [comment]
Thanks for your patience and efforts to help me! this is really driving me nuts 

<img width="808" alt="Screen Shot 2019-09-03 at 1 56 14 PM" src="https://user-images.githubusercontent.com/7185059/64196798-a0af6480-ce52-11e9-9349-2f064f325891.png">

**Comment by cwilby** [comment]
No worries, just one last sanity check - do you have `__REDIS_PASSWORD__` in `redis.conf`?

**Comment by javierrivarola** [comment]
yes! if i login into the container i can auth to redis

<img width="269" alt="Screen Shot 2019-09-03 at 2 06 45 PM" src="https://user-images.githubusercontent.com/7185059/64197420-1d8f0e00-ce54-11e9-82ab-46d09e3773f2.png">

**Comment by cwilby** [comment]
Ok, so the password is definetely being set :+1: 

Are your workspace/php-fpm instances connected to the `backend` network?

**Comment by javierrivarola** [comment]
yeah, if i remove the password everything works

**Comment by javierrivarola** [comment]
Shout out to @cwilby who helped me resolve this issue!! He is the real MVP guys! 
The last step missing for me was removing the bind directive in redis.conf file, and everything worked!

**Comment by quangvision** [comment]
Hi
I want to add up one reason behind "Connection refused to Redis server" is your Laravel app may still caching .env configuration file so it is good to remove all config cache by running `php artisan cache:clear` and `php artisan config:clear` to reflect new .env file with REDIS_PASSWORD set :)

**Comment by bumbummen99** [comment]
When experiencing strange issues with Laravel it is always a good idea to consider clearing caches, especially for config and views.

**Comment by cwilby** [comment]
Should probably put this somewhere else, but here's my modification of `workspace/docker-entrypoint.sh` that helps me avoid cache during dev.

```
#!/bin/bash
set -e

while ! mysqladmin ping -h"${DB_HOST}" -u"root" -p"${ROOT_PASSWORD}" --silent; do
  sleep 1
done

pushd /var/www
  if [ "${APP_ENV}" = "production" ]; then
    sed -i 's|APP_DEBUG=true|APP_DEBUG=false|g' /var/www/.env
    composer install --no-plugins --no-scripts --no-dev
  else
    composer install --no-plugins --no-scripts
  fi

  [ -z "${APP_KEY}" ] && php artisan key:generate --force
  [ -z "${JWT_SECRET}" ] && php artisan jwt:secret --force

  php artisan migrate --force
  php artisan db:seed --force

  if [ "${APP_ENV}" = "production" ]; then
    php artisan clear-compiled
    php artisan cache:clear
    php artisan config:clear
    php artisan clockwork:clean
    php artisan event:clear
    php artisan optimize:clear
    php artisan route:clear
    php artisan view:clear

    php artisan config:cache
    php artisan event:cache
    php artisan route:cache
    php artisan view:cache
  fi
  
popd

/sbin/my_init
```

**Comment by bumbummen99** [comment]
@cwilby That is really helpful, thanks alot! :+1:

**Comment by jrbecart** [comment]
> Shout out to @cwilby who helped me resolve this issue!! He is the real MVP guys!
> The last step missing for me was removing the bind directive in redis.conf file, and everything worked!

This step should be added in the documentation.

**Comment by dabarto** [comment]
> bind

can you detailed the solution ?

**Comment by jrbecart** [comment]
@dabarto 
In redis.conf comment out line 69:
```
# bind 127.0.0.1

**Comment by dabarto** [comment]
> # bind 127.0.0.1

after i comment the bind 127.0.0.1
please help

![image](https://user-images.githubusercontent.com/7286112/68479652-6bb1fa80-0265-11ea-88c1-dc7514d9040b.png)
![image](https://user-images.githubusercontent.com/7286112/68479660-74a2cc00-0265-11ea-9cbb-7fd7ec9e901c.png)
![image](https://user-images.githubusercontent.com/7286112/68479673-7b314380-0265-11ea-8d7e-f123faeacb42.png)

**Comment by cwilby** [comment]
@dabarto make sure:

- [x] `.env` has REDIS_PASSWORD
- [ ] `./laradock/redis/redis.conf` has lines `bind 127.0.0.1` and `requirepass __REDIS_PASSWORD`
- [ ] `./laradock/redis/Dockerfile` has `ARG REDIS_PASSWORD` and `sed` invocation
- [ ] `docker-compose.yml` has `REDIS_PASSWORD` build args

**Comment by moeinrahimi** [comment]
in docker-compose you can bind redis to localhost so it can't get accessed from outside
```
ports:
    - "127.0.0.1:27017:27017"
```
